### PR TITLE
Include Intune Enrollment ID with requests if present

### PR DESCRIPTION
**Pull request by midella**

### CODE REVIEW

File: ADAL/src/ADEnrollmentGateway.m [hunk 1]
```diff
@@ -146,6 +146,15 @@ + (NSString *)intuneMamResource:(NSString *)authority
     return resources[[ADEnrollmentGateway normalizeAuthority:authority]];
 }
 
++ (NSString *)intuneMAMResourceJSON:(NSString *)authority
+{
+    NSString* mamResource = [ADEnrollmentGateway intuneMamResource:authority];
+    mamResource = mamResource ? [NSString stringWithFormat:@"{%@:%@}",[ADEnrollmentGateway normalizeAuthority:authority],mamResource] : nil ;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: nit:space after comma

File: ADAL/src/ADEnrollmentGateway.h [hunk 2]
```diff
@@ -29,6 +29,8 @@
 + (NSString *)allEnrollmentIds;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.h]: I am thinking that, should we rename it `allEnrollmentIdsJSON`? indicating it is returning a JSON blob. What do you think? :)

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 3]
```diff
@@ -126,6 +127,13 @@ - (void)acquireTokenByRefreshToken:(NSString*)refreshToken
         request_data[OAUTH2_SCOPE] = _requestParams.scope;
     }
     
+
+    NSString* enrollId = [ADEnrollmentGateway enrollmentIDForTokenUserID:(cacheItem ? (cacheItem.userInformation ? cacheItem.userInformation.userId : nil) : nil)
+                                                           requestUserID:_requestParams.identifier ? _requestParams.identifier.userId : nil];
+
```
Comment (jasoncoolmax) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: ADAL finds `cacheItem` by `_requestParams.identifier.userId `, which means `cacheItem.userInformation.userId` will always be the same as `_requestParams.identifier.userId`.
So for silent flow here, checking `cacheItem.userInformation.userId` should be enough.

I guess the fallback for silent flow is:
look up by unique account id (`cacheItem.userInformation.homeUserId`) 
-> look up by user id (`cacheItem.userInformation.userId`) 
-> return whatever we have

` cacheItem.userInformation.homeUserId` property is only available in dev branch, therefore we just pass in nil for 2.6.4 branch:
`[ADEnrollmentGateway enrollmentIDForUniqueAccountID:nil userId:cacheItem.userInformation.userId];`

What do you think?

Comment (midella) [reply ADAL/src/request/ADAcquireTokenSilentHandler.m]: I agree, I've made the changes.  For acquireTokenSilent, I've used the userId from _requestParams first in case there is no cacheItem

File: ADAL/src/ADEnrollmentGateway.m [hunk 4]
```diff
@@ -109,6 +117,29 @@ + (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId
...
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * __unused dic) {
+        return true;
+    }];
+}
+
++ (NSString*)enrollmentIDForHomeUserID:(NSString*) homeUserID userID:(NSString*) userID
+{
+    NSString* enrollmentID;
+    enrollmentID = homeUserID ? [ADEnrollmentGateway enrollmentIdForUserId:homeUserID] : nil;
+    if (enrollmentID)
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: i guess we should use `[ADEnrollmentGateway enrollmentIdForUniqueAccountId:homeUserId];` for `homeUserId`. 
`UniqueAccountId` and `homeUserId` are actually referring to the same thing. Sorry that it causes confusion to you. I will talk to Olga see if we could rename `homeUserId` in cache item to `UniqueAccountId`:)

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 5]
```diff
@@ -126,6 +127,12 @@ - (void)acquireTokenByRefreshToken:(NSString*)refreshToken
         request_data[OAUTH2_SCOPE] = _requestParams.scope;
     }
     
+
+    NSString* enrollId = [ADEnrollmentGateway enrollmentIDForHomeUserID:nil userID:(_requestParams.identifier ? _requestParams.identifier.userId : nil)];
```
Comment (jasoncoolmax) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: User information in token item is more reliable, as they are sent back from server. 
So i am thinking if we could make it 
`NSString* enrollId = [ADEnrollmentGateway enrollmentIDForHomeUserID:nil userID:(cacheItem.userInformation.userId ? cacheItem.userInformation.userId : _requestParams.identifier.userId)];`

File: ADAL/src/ADEnrollmentGateway.m (hunk 4 above)
Comment (jasoncoolmax) [reply ADAL/src/ADEnrollmentGateway.m]: Just talk to Olga, we are going to to rename both `uniqueAccountId` and `homeUserId` to `homeAccountId` :)

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m [hunk 6]
```diff
@@ -514,7 +515,12 @@ - (void)requestTokenByCode:(NSString *)code
     {
         [request_data setValue:_requestParams.scope forKey:OAUTH2_SCOPE];
     }
-    
+
+    NSString* enrollId = [ADEnrollmentGateway enrollmentIDForHomeUserID:nil userID:(_requestParams.identifier ? _requestParams.identifier.userId : nil)];
```
Comment (jasoncoolmax) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: nit: Doesn’t calling [ADEnrollmentGateway enrollmentIDForHomeUserID:nil userID: _requestParams.identifier.userId] have the same effect? :)

File: ADAL/src/ADEnrollmentGateway.m [hunk 7]
```diff
@@ -110,71 +116,114 @@ + (NSString *)allEnrollmentIds
...
+    // this will just return the first enrollment ID
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * __unused dic) {
+        return true;
     }
+                                                       error:error];
+}
+
++ (NSString*)enrollmentIDForUniqueAccountID:(NSString*) homeUserID userID:(NSString*) userID error:(ADAuthenticationError*__autoreleasing *) error
+{
+    NSString* enrollmentID;
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: I think this logic should be following (@jasoncoolmax, please double check):
`
if (homeAccountID)
{
    // If homeAccountID is provided, always require an exact match
    return [ADEnrollmentGateway enrollmentIdForUniqueAccountId:homeUserID error:error];
}
else
{
    NSString *enrollmentID = nil;

    // If legacy userID is provided and we didn't find an exact match, do a fallback to any enrollment ID to support no userID or single userID scenarios
    if (userID) enrollmentID = [ADEnrollmentGateway enrollmentIdForUserId:userID error:error];
    if (enrollmentID) return enrollmentID;

    return ADEnrollmentGateway enrollmentIdIfAvailable:error];
}

`

Basically, I don't think we need a fallback if homeAccountId was provided.

Comment (jasoncoolmax) [reply ADAL/src/ADEnrollmentGateway.m]: I think the above solution makes sense to me. 

I want to call out that, after server side make the changes (returning us user info together with auth code), we will remove all fallback in this function and only have exact match :)

File: ADAL/src/ADEnrollmentGateway.m [hunk 8]
```diff
@@ -110,71 +125,120 @@ + (NSString *)allEnrollmentIds
...
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * __unused dic) {
+        return true;
     }
+                                                       error:error];
+}
+
++ (NSString *)enrollmentIDForUniqueAccountID:(NSString *) homeUserID userID:(NSString *) userID error:(ADAuthenticationError *__autoreleasing *) error
+{
+    NSString *enrollmentID;
+    enrollmentID = homeUserID ? [ADEnrollmentGateway enrollmentIdForUniqueAccountId:homeUserID error:error] : nil;
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: please see my previous comment about logic update for this method (it got collapsed with code style updates, so just making sure to address it)

File: ADAL/src/ADEnrollmentGateway.m [hunk 9]
```diff
@@ -110,71 +125,120 @@ + (NSString *)allEnrollmentIds
...
+        return enrollmentID;
+    
+    enrollmentID = [ADEnrollmentGateway enrollmentIdIfAvailable:error];
+    return enrollmentID;
+}
+
++ (NSString *)intuneMAMResourceJSON:(NSString *)authority error:(ADAuthenticationError *__autoreleasing *)error
+{
+    NSString *mamResource = [ADEnrollmentGateway intuneMamResource:authority error:error];
+    NSString *normalizedAuthority = [[NSURL URLWithString:authority] adHostWithPortIfNecessary];
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: nit: this is technically not authority, can we rename this variable to host/normalizedHost for clarity?

File: ADAL/src/ADEnrollmentGateway.m [hunk 10]
```diff
@@ -110,71 +125,120 @@ + (NSString *)allEnrollmentIds
...
         return nil;
     }
     else if (![resources isKindOfClass:[NSDictionary class]])
     {
-        AD_LOG_WARN(nil, @"Intune Resource JSON structure is incorrect. (Not a dictionary)");
+        if(error)
+        {
+            *error = [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_DEVELOPER_JSON_DATA_STRUCTURE_MALFORMED
+                                                                  protocolCode:nil
+                                                                  errorDetails:@"Intune Resource JSON structure is incorrect. (Not a dictionary)"
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: 👍

File: ADAL/src/ADEnrollmentGateway.m [hunk 11]
```diff
@@ -110,71 +125,120 @@ + (NSString *)allEnrollmentIds
...
+                                                                 correlationId:nil];
+        }
         return nil;
     }
 
-    NSString* normalizedAuthority = [[NSURL URLWithString:authority] adHostWithPortIfNecessary];
-    return resources[normalizedAuthority];
+    NSArray<NSURL *> *aliases = [[ADAuthorityValidation sharedInstance] cacheAliasesForAuthority:[NSURL URLWithString:authority]];
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: nit: why don't we pass authority as URL in the first place?

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 12]
```diff
@@ -126,6 +127,14 @@ - (void)acquireTokenByRefreshToken:(NSString*)refreshToken
         request_data[OAUTH2_SCOPE] = _requestParams.scope;
     }
     
+    if (![ADHelpers isADFSInstance:_requestParams.authority])
+    {
+        NSString* enrollId = [ADEnrollmentGateway enrollmentIDForUniqueAccountID:nil userID:(cacheItem.userInformation.userId ? cacheItem.userInformation.userId : _requestParams.identifier.userId) error:NULL];
```
Comment (oldalton) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: very nit: this conditional is a bit long to read in a combination with the method call :)

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 13]
```diff
@@ -126,6 +127,14 @@ - (void)acquireTokenByRefreshToken:(NSString*)refreshToken
...
+
+        if (enrollId)
+            [request_data setObject:enrollId forKey:@"microsoft_enrollment_id"];
```
Comment (oldalton) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: nit: can we declare "microsoft_enrollment_id" const in Oauth2 constants?

File: ADAL/src/request/ADAcquireTokenSilentHandler.m (hunk 12 above)
Comment (oldalton) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: why are we ignoring the error?

File: ADAL/src/request/ADAuthenticationRequest+Broker.m [hunk 14]
```diff
@@ -317,6 +321,8 @@ - (NSURL *)composeBrokerRequest:(ADAuthenticationError* __autoreleasing *)error
       BROKER_MAX_PROTOCOL_VERSION : @"2",
       @"extra_qp"       : _queryParams ? _queryParams : @"",
       @"claims"         : _claims ? _claims : @"",
+      @"enrollment_ids" : enrollmentIds ? enrollmentIds : @"",
```
Comment (oldalton) [inline ADAL/src/request/ADAuthenticationRequest+Broker.m]: nit: @jasoncoolmax can we call "enrollment_ids" "intune_enrollment_ids" or "ca_enrollment_ids" here? If we were to publish the broker protocol, what would be easier to understand?

File: ADAL/tests/XCTestCase+TestHelperMethods.m [hunk 15]
```diff
@@ -49,6 +50,9 @@ @implementation XCTestCase (TestHelperMethods)
 
 volatile int sAsyncExecuted;//The number of asynchronous callbacks executed.
 
+IMP originalAllEnrollmentIds;
```
Comment (oldalton) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: is this used?

File: ADAL/src/request/ADAcquireTokenSilentHandler.m (hunk 12 above)
Comment (jasoncoolmax) [reply ADAL/src/request/ADAcquireTokenSilentHandler.m]: Although we are not going to stop the whole acquire token flow because of this error, I think we could log it here. It will be helpful diagnosis purpose :)

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m [hunk 16]
```diff
@@ -514,7 +515,16 @@ - (void)requestTokenByCode:(NSString *)code
     {
         [request_data setValue:_requestParams.scope forKey:OAUTH2_SCOPE];
     }
-    
+
+    if (![ADHelpers isADFSInstance:_requestParams.authority])
+    {
+        NSString* enrollId = [ADEnrollmentGateway enrollmentIDForUniqueAccountID:nil userID:_requestParams.identifier.userId error:NULL];
```
Comment (jasoncoolmax) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: Same here, probably we could log the error if any :)

File: ADAL/src/request/ADAuthenticationRequest+Broker.m (hunk 14 above)
Comment (jasoncoolmax) [reply ADAL/src/request/ADAuthenticationRequest+Broker.m]: I like "intune_enrollment_ids" better. 
But I think @midella may know better about which name might be more precise :)

File: ADAL/src/request/ADAcquireTokenSilentHandler.m (hunk 12 above)
Comment (midella) [reply ADAL/src/request/ADAcquireTokenSilentHandler.m]: This method originally only returned the error returned by NSJSONSerialization, and logged separately.  It's since been updated to accept ADAuthenticationErrors, and so passing the error here should trigger logging automatically.

File: ADAL/src/request/ADAuthenticationRequest+Broker.m (hunk 14 above)
Comment (midella) [reply ADAL/src/request/ADAuthenticationRequest+Broker.m]: intune_enrollment_ids is more precise

**Comment by jasoncoolmax** [comment]
same code style nit as the EnrollmentIDGateway PR, like  
`(ADAuthenticationError*__autoreleasing *) error` to
`(ADAuthenticationError * __autoreleasing *)error`
